### PR TITLE
Wraith rework

### PR DESCRIPTION
**Pull request by Sovexe**
[BALANCE] [FEATURE] [wiki]

forked from https://github.com/goonstation/goonstation/pull/8542 which is abandoned and not allowing commits

## About the PR 
This PR is an attempt at giving wraith more customizability and personnality while also providing fun counterplay!

### **Balance to existing wraith:**
-Point generation lowered accross the board.
-Removes the following abilities from the base wraith: "Make poltergeist" "Raise skeleton" "Make revenant"
-Introduces new rotating zones in which point generation is increased in order to invite the wraith to navigate throughout the station and not camp in one spot.
-Haunt is the same as before with one exception: Upon use turns the wraith visible, but no longer makes them incorporeal at the end of the ability's cooldown. Instead the wraith can go incorporeal again when they hit "haunt" once more when they want it.

### **Additions**
-Adds three new wraith paths, each with their own abilities and weaknesses. You can transform into a wraith path once enough bodies have been consumed.

image

### _Path 1: Harbinger_
Meant as a summoner. Possesses the following abilities:

-"**Raise skeleton**" which now also works on closets. There is a maximum of one skeleton per closet.

-"**Make revenant**"

-"**Create void portal**" which creates a summoning circle that slowly grows in power and releases a small army of mobs the wraith can choose from a list. The stronger the portal, the more health it has, and the higher the mob cap/summon range. Destroying the portal or placing a new one removes all mobs that have been spawned by the portal.

-"**Summon void creature**" requests ghosts of players to join and offers a choice between three creatures to pick from, meant to aid the harbinger;

-Skeleton commander; A tanky mob with a long range stab, a sweeping strike, a raise minor skeleton ability, and a rally ability buffing nearby critters.
-Tentacle fiend; A squishy, ranged mob with a tentacle hook to bring people to them, and a "lash" usable on disabled people to keep them on the ground and cause a lot of bleeding damage. Meant to be used in conjunction with the horde the harbinger summons.
-Void hound; A fast, sneaky mob with a leap that stuns targets for a short time, a cloak making them semi-transparent, and a strong lacerating arm. Quite squishy and meant to be roaming from the back lines.

-"**Speak to summons**" which does what you'd expect.

### _Path 2: Plaguebringer_
Meant as a debuffer and poisoner. Possesses the following abilities:

-**"Curses"** which are targeted and can be used on humans. There are four curses each with their own small debuffs. Getting more curses gives increasing warnings to the target. Curses can be removed being splashed with holy water, or by getting a bible beating from a chaplain. You cannnot apply a curse to a target in the chapel.  Once a target has four curses on them, the "curse of death" is now usable on them, which kills the target and consumes its soul.
The curses are;
-Plague of blood; Gives the affected the hemophilia trait, makes them randomly puke blood.
-Plague of blindness. Regularly gives the owner eye damage.
-Plague of weakness. Regularly slows down the owner and reduces stamina regen as well as making them drowsy.
-Plague of rot. Makes the owner vomit at random intervals and makes them unable to eat any food.

-**"Poison"** can be used on a food item, glass or open container. Opens a list of poison and disease reagents, then adds that poison to the target.

-**"Summon rot hulk"** checks the area for gibs, vomit and other unpleasant things, then if the area is dirty enough begins to coalesce the filth into a rot hulk. The rot hulk is a mob with a simple ai, chasing people and attacking them, and exploding when killed or low health in a shower of poisoning gibs. The hulk is stronger if there is more filth around.

-**"Summon plague rat"** requests ghosts of players if they want to join as a plague rat. The plague rat is a small, squishy mob with two abilities allowing it to bite someone and inject them with a very low dose of "rat spit", and another ability allowing them to eat filth and gibs. Rat spit is essentially a weaker "rat venom" that gives you the "rat plague" which in turn makes you contract more diseases.
If they eat enough they go up in size and gain the "summon warren" ability. There can be one warren per plague rat, and each warren slowly pours out aggressive mice out of it, up to a cap, and very slowly heals neighbouring plague rats. 
If they grow again, they can no longer pass doors and tables, gain a strong arm, and gain the "slam" ability as well as the "summon plague rat" ability like the wraith's. There is a limit to the amount of plague rats on the map.

-"**Speak to summons**" 

### _Path 3: Trickster_
Meants as a sneaky, confusing liar. Possesses the following abilities:

-**"Make poltergeist"**

-**"Mass whisper"** whispers to everyone in the area

-**"Fake sound"** Play a fake sound from a list at the target area.

-**"Creeping dread"** Causes the target to contract a debuff that ticks down while near lights, and up otherwise. As the debuff grows in power while in the dark, the target begins to shake, hallucinate, take minor damage and eventually reaches a point where it developps into a heart attack.

-**"Hallucinate"** causes the living target to hallucinate for awhile, with all new hallucinations such as fake announcements, blood pourring out of the walls, and more!

-**"Lay rune trap"** is an ability for which the trickster must manifest. Lets them plant a trap that becomes invisible if no lit lights are close to it. The trap triggers when a human runs on it, but wont trigger if they walk/crawl. It can be destroyed with any item. When sprung, the trap will have a variety of effect depending on the choice the wraith made when laying it down. Effects range from minor explosions, fire, emps and more!

-**"Haunt"** is now different for this wraith, allowing it to choose a human target beforehand, and copying when haunting. The target can be alive, dead, missing any number of limbs/head, or any mutantrace. Being in proximity to people while haunting grants the wraith "possession points" used by the "possession" ability.

-**"Possession"** allows the wraith, after a long delay and if their possession points are high enough, to take control of a living human. The control lasts for about a minute, then returns the wraith to the afterlife when finished, returning the human to their body with non-negligible brain damage.

### **Additional counterplay**

A new item has been introduced: The spirit candle. It can simply be made by combining a skull and a candle, then can be lit with an open flame. The candle lasts about 2 minutes and cannot be reused once it's expired. While lit, the candle emits a distinct blue glow and will reveal the wraith when it is nearby. Using the candle in hand will snuff it out, darken the area around, and force the wraith to manifest for thirty seconds, similarly to how salt works. Meant to be used as a more "proactive" play against the wraith.

skullcandle

Each wraith path is also countered in a lesser way through actions the crew can take:
Plaguebringer's influence can be lessened by keeping the place clean and passing out holy water.
Trickster's abilities are weaker if the crew replaces the lights and keeps an eye out for impersonators and mind tricks.
Harbinger's damage is mostly physical, and requires bodies which can be disposed of quickly to prevent their stronger abilities.

To-do list:

- [x] Finish all lines and tasks with "Todo" in the code.
- [x] Sprites
- [x] Sound effects
- [x] Animations and visual effects
- [x] Add more unique wraith vulnerabilities and counterplay
- [x] Fix a bug causing the rot hulk's ai to hang
- [x] Improve trickster's haunt ability to allow it to move a bit faster and open non-id locked doors.
 
Icon showcase:

image
abilshowcase
objshowcase

## Why's this needed? 
Provides a greater degree of customizability and keeps the game fresh.
Allows the wraith player to choose the approach they like most.
Provides more engaging counterplay besides simply salting everything everywhere.

## Changelog

```changelog
(u)Bartimeus973
(*)Overhaul to the wraith. Wraiths can now pick from three paths upon consuming enough bodies, each with their unique abilities. Sprites done in collaboration with: Swinglow, CalliopeSoups, Mora and Shiiba!
(*)Adds a new item to counter wraith: the spirit candle. Created by combining a skull and a candle. While lit in hand it reveals wraiths in the surrounding area and using it in hand will manifest the wraith. One time use.
```

### CODE REVIEW

**Comment by nefarious6th** [comment]
looks good and will put an end to chain-revenanting destructo-wraiths while providing good counterplay. I predict the harbinger will probably be the most popular wraith form, but I'd be excited to see them all at some point in a live setting even if it was just for testing.

File: code/obj/machinery/vortex_wraith.dm [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+					src.total_mob_value -= point_value
+					critter_list -= M
+
+			if (growth > 0)	//Are we spawning mobs yet? If not, wait for next tick
+				var/list/eligible_turf = list()
+				var/turf/chosen_turf = null
+				for (var/mob/living/carbon/human/H in range((1 + src.growth), src))	//Lets try to spawn near someone
+					if(isalive(H))
+						for (var/turf/T in range(3, H))
+							if (isturf(T) && istype(T, /turf/simulated/floor))	//Find a non-wall, non-space turf to spawn in
```
Comment (ZeWaka) [inline code/obj/machinery/vortex_wraith.dm:73]:
```suggestion
							if (istype(T, /turf/simulated/floor))	//Find a non-wall, non-space turf to spawn in
```

File: code/obj/machinery/vortex_wraith.dm [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+								eligible_turf += T
+				if (length(eligible_turf) <= 0)	//No spot to spawn near a human, or no human in range, lets try to find a regular turf instead
+					for (var/turf/T in range((1 + src.growth), src))
+						if (isturf(T) && istype(T, /turf/simulated/floor))	//Find a non-wall, non-space turf to spawn in
```
Comment (ZeWaka) [inline code/obj/machinery/vortex_wraith.dm:77]:
```suggestion
						if (istype(T, /turf/simulated/floor))	//Find a non-wall, non-space turf to spawn in
```

File: code/obj/machinery/runetrap.dm [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+/obj/machinery/wraith/runetrap/explosive
+	Crossed(atom/movable/A)
+		..()
+		if(!src.armed)
+			return 1
+		if(ismob(A) && !istype(A, /mob/living/critter/wraith/trickster_puppet) && !istype(A, /mob/wraith))
+			var/mob/M = A
+			if(checkRun(M))
+				src.visible_message("<span class='alert>[M] steps on [src] and triggers it! You hear a buzzing sound!</span>")
+				playsound(src, "sound/voice/wraith/wraithraise3.ogg", 80)
```
Comment (ZeWaka) [inline code/obj/machinery/runetrap.dm:194]:
```suggestion
				playsound(src, 'sound/voice/wraith/wraithraise3.ogg', 80)
```

File: code/obj/machinery/runetrap.dm [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+			var/mob/M = A
+			if(checkRun(M))
+				src.visible_message("<span class='alert>[M] steps on [src] and triggers it! You can hear a slippery sound!</span>")
+				M.remove_pulling()
+				M.changeStatus("weakened", 3 SECONDS)
+				boutput(M, "<span class='notice'>You suddenly slip!</span>")
+				playsound(M, "sound/misc/slip.ogg", 50, 1, -3)
+				var/atom/target = get_edge_target_turf(M, M.dir)
+				M.throw_at(target, 12, 1, throw_type = THROW_SLIP)
+				playsound(src, "sound/voice/wraith/wraithraise3.ogg", 80)
```
Comment (ZeWaka) [inline code/obj/machinery/runetrap.dm:214]:
```suggestion
				boutput(M, "<span class='notice'>You suddenly slip!</span>")
				playsound(M, 'sound/misc/slip.ogg', 50, 1, -3)
				var/atom/target = get_edge_target_turf(M, M.dir)
				M.throw_at(target, 12, 1, throw_type = THROW_SLIP)
				playsound(src, 'sound/voice/wraith/wraithraise3.ogg', 80)
```
i imagine every sound in this PR is using the old incorrect way

File: code/modules/food_and_drink/ingredients.dm [hunk 5]
```diff
@@ -942,3 +963,10 @@ obj/item/reagent_containers/food/snacks/ingredient/pepperoni_log
 	initial_volume = 15
 	initial_reagents = list("mercury"=1,"LSD"=1,"ethanol"=1,"gcheese"=1)
 	food_effects = list("food_sweaty","food_bad_breath")
+
+	heal(var/mob/M)
+		if (istype(M, /mob/living/critter/wraith/plaguerat))
+			boutput(M, "<span class='notice'>This is incredible, but there isnt enough! MORE!</span>")
+			M.reagents.add_reagent("Omnizine", 3)
+			M.reagents.add_reagent("methamphetamine", 3)
```
Comment (ZeWaka) [inline code/modules/food_and_drink/ingredients.dm:971]: why random capitals

File: code/mob/living/critter/wraith/spiker.dm [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+	on_launch(var/obj/projectile/P)
+		..()
+		if (!("owner" in P.special_data))
+			P.die()
+			return
+
+	on_end(var/obj/projectile/P)	//Clean up behind us
+		SPAWN(1 DECI SECOND)
+		if (previous_line != null)
```
Comment (ZeWaka) [inline code/mob/living/critter/wraith/spiker.dm:91]: so much cargo cult useless code

File: code/modules/status_system/wraithEffects.dm [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+		H = owner
+		get_image_group(CLIENT_IMAGE_GROUP_ILLUSSION).remove_mob(H)
+
+	onUpdate()
+		var/mult = 1
+		if (probmult(5))
+			switch (rand(1,3))
+				if (1) // Image based illusion
+					for (var/turf/W in range(6, H))	//Check for surrounding spots
+						wall_list += W
```
Comment (ZeWaka) [inline code/modules/status_system/wraithEffects.dm:148]: should be a block(), and isn't even checking for walls lol, so much code like this

File: code/obj/machinery/runetrap.dm [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+/obj/machinery/wraith/runetrap/emp
+	Crossed(atom/movable/A)
+		..()
+		if(!src.armed)
+			return 1
+		if(ismob(A) && !istype(A, /mob/living/critter/wraith/trickster_puppet) && !istype(A, /mob/wraith))
+			var/mob/M = A
+			if(checkRun(M))
+				var/turf/T = get_turf(M)
+				for (var/atom/O in T.contents)
```
Comment (ZeWaka) [inline code/obj/machinery/runetrap.dm:116]: yes, `O`, my favorite atom name

File: code/obj/machinery/rat_den.dm [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+					M.linked_den = src
+					linked_critters ++
+
+		//Plague rats in range heal up slowly
+		for (var/mob/living/critter/wraith/plaguerat/P in range(5, src))
+			if((P.health < (P.health_brute + P.health_burn)))
+				for(var/damage_type in P.healthlist)
+					var/datum/healthHolder/hh = P.healthlist[damage_type]
+					hh.HealDamage(3)
+				boutput(P, "The proximity of the rat den fills you with renewed malevolence.")
```
Comment (ZeWaka) [inline code/obj/machinery/rat_den.dm:45]: tons of 5+ range checks in this PR - stuff like this should really be a by_type[] and dist check - would say spatial hashing but that's only clients

File: code/modules/status_system/statusEffects.dm [hunk 10]
```diff
@@ -2169,3 +2178,48 @@
...
+	onRemove()
+		var/mob/M = owner
+		if (istype(M, /mob/wraith))
+			var/mob/wraith/W = M
+			W.forced_manifest = FALSE
+			W.haunting = FALSE
+			W.flags |= UNCRUSHABLE
+			if (!istype_exact (M, /mob/wraith/poltergeist))
+				M.alpha = 160
+		else if (istype_exact (M, /mob/wraith/poltergeist))
```
Comment (ZeWaka) [inline code/modules/status_system/statusEffects.dm:2218]: pls
```suggestion
			if (!istype_exact(M, /mob/wraith/poltergeist))
				M.alpha = 160
		else if (istype_exact(M, /mob/wraith/poltergeist))
```

File: code/modules/medical/genetics/bioEffects/non_genetic.dm [hunk 11]
```diff
@@ -271,3 +271,130 @@
...
+	acceptable_in_mutini = 0
+	probability = 0
+	curable_by_mutadone = 0
+
+	OnAdd()
+		if (ishuman(owner))
+			owner.traitHolder?.addTrait("hemophilia")
+
+	OnLife()
+		var/mult = 1
```
Comment (ZeWaka) [inline code/modules/medical/genetics/bioEffects/non_genetic.dm:296]: i like how mult is just always 1

just avoiding the actual problem lol

File: code/modules/medical/genetics/bioEffects/non_genetic.dm [hunk 12]
```diff
@@ -271,3 +271,130 @@
...
+	effectType = EFFECT_TYPE_DISABILITY
+	can_copy = 0
+	isBad = 1
+	occur_in_genepools = 0
+	acceptable_in_mutini = 0
+	probability = 0
+	curable_by_mutadone = 0
+
+	OnLife()
+		var/mult = 1
```
Comment (ZeWaka) [inline code/modules/medical/genetics/bioEffects/non_genetic.dm:325]: ditto

File: code/mob/wraith.dm [hunk 13]
```diff
@@ -639,6 +710,58 @@
...
+/mob/wraith/wraith_decay
+	name = "Plaguebringer"
+	real_name = "plaguebringer"
+	desc = "A pestilent ghost, spreading disease wherever it goes. Just looking at it makes you queasy."
+	icon = 'icons/mob/mob.dmi'
+	icon_state = "wraith_plague"
+
+	New(var/mob/M)
+		..()
+		src.addAllDecayAbilities()
```
Comment (ZeWaka) [inline code/mob/wraith.dm:726]: but why make a whole new proc for this

File: code/mob/wraith.dm [hunk 14]
```diff
@@ -639,6 +710,58 @@
...
+/mob/wraith/wraith_trickster
+	name = "trickster"
+	real_name = "trickster"
+	desc = "A living shadow seeking to disrupt the station with lies and deception."
+	icon = 'icons/mob/mob.dmi'
+	icon_state = "wraith_trickster"
+	var/points_to_possess = 50	//How many points do we need to possess someone?
+	var/possession_points = 0	//How many do we currently have?
+	var/mutable_appearance/copied_appearance = null	//Steal someone's appearance and use it during haunt
+	var/copied_desc = null	//Steal their descriptions too
```
Comment (ZeWaka) [inline code/mob/wraith.dm:750]: code doc 😩

File: code/mob/wraith.dm [hunk 15]
```diff
@@ -591,7 +618,51 @@
...
+			var/list/safe_area_names = list()
+			var/num_safe_areas = clamp(3, 1, 5)
+			var/area/temp = null
+			var/list/locations_copy = list()
+			var/has_client = false
+			for(var/A in valid_locations)
+				for(var/mob/living/carbon/human/M in A)	//Check for player_controlled mobs in the zone
+					if(M.client != null)
+						has_client = true
+						continue
```
Comment (ZeWaka) [inline code/mob/wraith.dm:653]: not a break??

File: code/mob/wraith.dm [hunk 16]
```diff
@@ -591,7 +618,51 @@
...
+				if (has_client)
+					locations_copy.Add(A)
+					has_client = false
+			if (length(locations_copy) < 3) //Not enough people?
+				for(var/A in valid_locations)	//Get random available areas instead
+					locations_copy.Add(A)
+			for(var/i = 0, i < num_safe_areas, i++)
```
Comment (ZeWaka) [inline code/mob/wraith.dm:660]: why this type of loop

File: code/mob/wraith.dm [hunk 17]
```diff
@@ -591,7 +618,51 @@
...
+
+		get_new_booster_zones()	//Get new zones you can get more power in.
+			booster_locations = list()
+			var/list/safe_area_names = list()
+			var/num_safe_areas = clamp(3, 1, 5)
+			var/area/temp = null
+			var/list/locations_copy = list()
+			var/has_client = false
+			for(var/A in valid_locations)
+				for(var/mob/living/carbon/human/M in A)	//Check for player_controlled mobs in the zone
```
Comment (ZeWaka) [inline code/mob/wraith.dm:650]: this is !very! expensive

File: code/mob/wraith.dm [hunk 18]
```diff
@@ -138,24 +157,43 @@
...
 		if (src.haunting)
-			src.hauntBonus = 0
 			for (var/mob/living/carbon/human/H in viewers(6, src))
 				if (!H.stat && !H.bioHolder.HasEffect("revenant"))
 					src.hauntBonus += 5
+					if(istype(src, /mob/wraith/wraith_trickster))
+						src.hauntBonus += 1
+
+		for (var/obj/machinery/wraith/vortex_wraith/V in range(8, src))
+			if (V == linked_portal)
```
Comment (ZeWaka) [inline code/mob/wraith.dm:172]: another by_type, like come on it's range 8!

**Review by ZeWaka — CHANGES_REQUESTED** [review]
i also wanted a bunch of this code to not be scattered in 30 different places, and to live in its own  `modules/gamemode/wraith` folder
this is just as much as i could review in 15m

File: code/modules/medical/genetics/bioEffects/non_genetic.dm (hunk 11 above)
Comment (Sovexe) [reply code/modules/medical/genetics/bioEffects/non_genetic.dm:296]: What makes this odd to me is that OnLife() actually gets passed a mult var already

**Pull request merged**